dartagnan/src/main/java/com/dat3m/dartagnan/wmm/relation/base/stat/StaticRelation.java

```diff
@@ -47,4 +55,35 @@ protected BooleanFormula encodeApprox(SolverContext ctx) {
         }
         return enc;
     }
+
+    protected void addMatchingTupleSet(List<Event> events, String open, String close, Consumer<Tuple> outMust) {
```

Comment:
The naming of the method is bad. I would call it `findMatchingPairs` and rename the parameters to something like `tagA` and `tagB` (or even use FIlters as input).
Also, you can call the consumer `mustMatch` and additionally add a `mayMatch` consumer.
The `mayMatch` consumer can be used to populate `maxTupleSet`.
I think this makes this method a lot more generic and reusable**.
That being said, this method is not usable for `RelCrit`, so it has a single valid user right now. You might as well put it into `RelRMW` as a helper method.

** If you want to go all the way, you could even add an option to change between "nested matching" (used by `RelCrit`) and "linear/direct matching" (used by LL/SC). In this case, it might even be worth to create a class that encapsulates these features.

Reply:
I decided to inline the method again, since both usages were indeed incompatible.  Feel free to outline it again, if the need arises.

Reply:
I agree with the inlining. If there is only one user, there is not much advantage in having a separate method (the only one is that it possibly has a descriptive name).